File: 140.word-break-ii.js
/*
 * @lc app=leetcode id=140 lang=javascript
 *
 * [140] Word Break II
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {string[]}
 */
var wordBreak = function (s, wordDict) {
  wordDict = new Set(wordDict);
  const ans = [];
  const temp = [];
  const DFS = (start) => {
    if (start === s.length) {
      ans.push(temp.join(" "));
      return;
    }
    for (let i = start; i < s.length; ++i) {
      const word = s.slice(start, i + 1);
      if (wordDict.has(word)) {
        temp.push(word);
        DFS(i + 1);
        temp.pop();
      }
    }
  };

  DFS(0);
  return ans;
};
// @lc code=end

File: solutions/140.word-break-ii.rb
# @param {String} s
# @param {String[]} word_dict
# @return {String[]}
def word_break(s, word_dict)
  res = []
  tmp = []
  dfs = lambda do |starter|
    res << tmp.join(' ') if starter == s.size
    (starter...s.size).each do |ender|
      sub_str = s[starter..ender]
      if word_dict.include?(sub_str)
        tmp << sub_str
        dfs.call(ender+1)
        tmp.pop
      end
    end
  end
  dfs.call(0)
  res
end
